lib/puma/launcher.rb

```diff
@@ -297,8 +297,10 @@ def prune_bundler
 
       log '* Pruning Bundler environment'
       home = ENV['GEM_HOME']
+      bundle_gemfile = ENV['BUNDLE_GEMFILE']
```

Comment:
Oh, what happened?

Reply:
Something about the environment in the Travis build was causing `Bundler.with_original_env` to make Puma load in an endless loop. Puma would load, it would call `Bundler.with_original_env`, then call `Kernel.exec('puma-wild', ...)`, then the loop would start over again.

You can see that loop in [this failed build](https://travis-ci.org/puma/puma/jobs/650170319?utm_medium=notification&utm_source=github_status).

I could not reproduce this issue locally, even with the same version of Ubuntu, same version of Ruby, and same version of RVM. There are probably other pieces in the environment on the Travis build that I am not aware of and did not fully replicate.

Testing this required, as you can see, pushing many commits to track down the problem.

The `prune_bundler` method is recursive, as you know, and it relies on the line `return unless defined?(Bundler)` as its terminal case. Something about the Travis environment (I suspect it's RVM) was causing Bundler to be loaded every time the method recursed.

As I am out of time to work on this, I went back to the hacky solution that copies the needed env var (`BUNDLE_GEMFILE`) back into the environment after we run `Bundler.with_clean_env`.

It's not my favorite solution, at all. But it meets our needs and should not break anything in other people's environments.

Reply:
> I could not reproduce this issue locally, even with the same version of Ubuntu, same version of Ruby, same version of Bundler, and same version of RVM. There are probably other pieces in the environment on the Travis build that I am not aware of and did not fully replicate.

Travis is pretty dirty in my experience, so I've switched my repos to other CIs. You can consider something similar too.

Reply:
We've moved almost everything to Github Actions, but Ruby 2.2 can't be moved (yet?).

Reply:
Ruby 2.2 is no more maintained, even security, but I understand that you want to support its users. But I personally consider its usage dangerous and prefer to drop support from gems.

Reply:
Since https://github.com/ruby/setup-ruby/releases/tag/v1.14.0 it is possible to use Ruby 2.2 on GitHub Actions... but it could be that there's something else missing for the Puma test setup (e.g. that version wasn't built with the OpenSSL version we want to test Puma with, I'm sure @MSP-Greg knows about this and maybe even have tried it, I don't have the time to double check it now)

Reply:
I should note that I believe Travis' environment actually saved us from shipping a bug here. *Something* in that particular environment was causing the Bundler constant to be loaded earlier than anticipated. I suspect it is some RVM magic.

Reply:
One difference between Travis and GitHub Actions is (was) the `bundle install` command

Travis does the following (https://travis-ci.org/puma/puma/jobs/656438353#L885)

    bundle install --jobs=3 --retry=3 --path=${BUNDLE_PATH:-vendor/bundle}

Before #2126 (5 days ago), `--path` wasn't used on GitHub Actions, now it is:

https://github.com/puma/puma/blob/8f282206dcdd07d765553179fe31d61acfd08dda/.github/workflows/ruby.yml#L38-L39

Haven't checked if it matters for this particular discussion but thought it was worth mentioning here